File: LeetCode/0225. Implement Stack using Queues.cpp
/*
0225. Implement Stack using Queues

使用队列去模拟栈操作。也就是说只能使用以下操作：
队尾添加元素，队首访问、取出元素，获取队列大小，查看队列是否为空。

在push时，每次插入，都将新元素添加到队尾，然后原队列元素用循环向后移动一位。这样新元素就在队首了。
*/

class MyStack {
public:
    queue<int> qStack;

    /** Initialize your data structure here. */
    MyStack() {

    }

    /** Push element x onto stack. */
    void push(int x) {
        // 每次插入，都将新元素添加到队尾，然后原队列元素用循环向后移动一位
        int qSize = qStack.size();
        qStack.push(x);
        while (qSize > 0) {
            int temp = qStack.front();
            qStack.pop();
            qStack.push(temp);
            qSize--;
        }
    }

    /** Removes the element on top of the stack and returns that element. */
    int pop() {
        int poppedElement = qStack.front();
        qStack.pop();
        return poppedElement;
    }

    /** Get the top element. */
    int top() {
        return qStack.front();
    }

    /** Returns whether the stack is empty. */
    bool empty() {
        return qStack.empty();
    }
};

File: LeetCode/0225. Implement Stack using Queues.py
'''
0225. Implement Stack using Queues

思路：用两个队列实现。
每次top或者pop，就转移到另一个队列中去，直到留下最后一个数，再pop或者top。

'''

class MyStack:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.q = [[], []]
        self.current = 0


    def push(self, x: int) -> None:
        """
        Push element x onto stack.
        """
        self.q[self.current].append(x)


    def pop(self) -> int:
        """
        Removes the element on top of the stack and returns that element.
        """
        while len(self.q[self.current]) > 1:
            self.q[1 - self.current].append(self.q[self.current].pop(0))
        top = self.q[self.current].pop(0)
        self.current = 1 - self.current
        return top


    def top(self) -> int:
        """
        Get the top element.
        """
        while len(self.q[self.current]) > 1:
            self.q[1 - self.current].append(self.q[self.current].pop(0))
        top = self.q[self.current].pop(0)
        self.q[1 - self.current].append(top)
        self.current = 1 - self.current
        return top



    def empty(self) -> bool:
        """
        Returns whether the stack is empty.
        """
        return len(self.q[self.current]) < 1



# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()